Mahmoud Ahmed Deok Ryong Kim

MahShaaban opened the pull request:
<!-- Please title your PR with all author's name -->
<!-- Two spaces at the end of a line = new line -->

**AUTHOR**

Dear @ReScience/editors,

I request a review for the following replication:

### Original article

**Title:**  Target analysis by integration of transcriptome and ChIP-seq data with BETA
**Author(s):**  S. Wang, H. Sun, J. Ma, C. Zang, C. Wang, J. Wang, Q. Tang, C. A. Meyer, Y. Zhang, and X. S. Liu.
**Journal (or Conference):**  Nature Protocols
**Year:**  2013
**DOI:**  10.1038/nprot.2013.150
**PDF:**  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4135175/pdf/nihms607566.pdf 

### Replication

**Author(s)**:   Mahmoud Ahmed and Deok Ryong Kim
**Repository**:  https://github.com/MahShaaban/ReScience-submission/tree/AHMED-KIM
**PDF**:  [AHMED-2019.pdf](https://github.com/ReScience/ReScience-submission/files/3660987/AHMED-2019.pdf)

**Keywords**:  gene-expression; transcription-factor; data-integration
**Language**:  R
**Domain**:  Computational Biology

### Results

* [x] Article has been fully replicated
* [ ] Article has been partially replicated
* [ ] Article has not been replicated


### Potential reviewers
<!-- If you know potential reviewers, you can tell us here -->
<!-- You can look at http://rescience.github.io/board for the -->
<!-- list of registered reviewers (but you can propose others) -->


---

**EDITOR**

* [ ] Editor acknowledgment
* [ ] Reviewer 1 
* [ ] Reviewer 2
* [ ] Review 1 decision [accept/reject]
* [ ] Review 2 decision [accept/reject]
* [ ] Editor decision [accept/reject]

MahShaaban committed:
remove article, add overleaf project as sumbodule article

MahShaaban committed:
updated code

MahShaaban committed:
add final pdf

MahShaaban committed:
Update README.md

rougier commented:
Hi @MahShaaban, 

Thank your for your submission and sorry for the delay in my answer.

There is actually a new procedure for submission which is supposed to be easier but since you started with the "old" one, we can continue using this PR (unless you want to switch). The idea was to avoid to first clone the submission repo but obviously you already did it. Just confirm it is fine to continue with this PR.

I'll assign an editor soon.

rougier commented:
@khinsen @ctb Can you edit this submission?

@MahShaaban Can you identify some potential reviewers in the board (http://rescience.github.io/board/)

khinsen commented:
@rougier This is a bit far from my competence. I suspect @ctb would be a better choice if he is available.

MahShaaban commented:
Thanks @rougier and @khinsen 

> There is actually a new procedure for submission which is supposed to be easier but since you started with the "old" one, we can continue using this PR (unless you want to switch). The idea was to avoid to first clone the submission repo but obviously you already did it. Just confirm it is fine to continue with this PR.

I am okay with this procedure
 
>  Can you identify some potential reviewers in the board (http://rescience.github.io/board/)

The link only lists the domain of expertise and the languages. Should I suggest reviewers based on that information or do I need to look up each reviewer for relevant expertise?

Thanks

rougier commented:
@ctb Would you be available to edit this submission ?
@MahShaaban you can try searching by keyword in the list of reviewer, but this i not mandatory at all. If @ctb accept, he'll find reviewers.

rougier commented:
@ReScience/editors We're looking for an editor for this submission.

rougier commented:
@MahShaaban , @delsuc will edit your submission.

delsuc commented:
@aaronshfman   could you review this submission ?

delsuc commented:
@aaronshifman  could you review this submission ?

delsuc commented:
@brunokieffer  could you review this submission ?

aaronshifman commented:
@delsuc I'd be happy to review this - I'm away for the next 2 weeks, but I can start after that

brunokieffer commented:
I will be happy to review it.

Sorry for this late answer, this procedure is a « première » for me.

Yours,

Le 8 oct. 2019 à 10:21, Marc-André Delsuc <notifications@github.com<mailto:notifications@github.com>> a écrit :


@brunokieffer<https://github.com/brunokieffer> could you review this submission ?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/ReScience/ReScience-submission/pull/59?email_source=notifications&email_token=ACCPUI2UTDSGCMYX7DF5E73QNQ7JLA5CNFSM4I3B7CNKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEATK5GY#issuecomment-539405979>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ACCPUI2AHMBVCDMFWVHPNR3QNQ7JLANCNFSM4I3B7CNA>.

____________________________________________________________________
Bruno KIEFFER

[cid:F5C62824-218B-4E1B-BD69-7F14ED299E84@u-strasbg.fr]


            IGBMC Biomolecular NMR group, CNRS UMR 7104

           ECOLE SUPERIEURE DE BIOTECHNOLOGIE DE STRASBOURG
             Boulevard Sebastien Brant  BP10413
             F-67412 ILLKIRCH Cedex    FRANCE

 Voice   : +33 (0)3 68 85 47 22
 Mobile : +33 (0)6 52 70 64 26
 FAX      : +33 (0)3 68 85 47 18

web : http://www-igbmc.u-strasbg.fr/research/department/3/team/35/

aaronshifman commented:
Apologies for the delay in getting this back.

The authors have implemented the BETA algorithm (binding and expression target analysis) for determining the role of transcription factors (inducer or repressor). The authors have claimed to fully replicate the original algorithm, and furthermore they analyse the sensitivity of the algorithm to changes in parameters. I have a few concerns about the implementation and the article which I've laid out here.

# Major Article Comments
* Unfortunately I came away from reading the introduction unclear as to how the algorithm works. Paragraph 2 ("Binding and Expression Target Analysis (BETA) [...]") outlines the steps the algorithm takes however many key terms seem to be undefined such as "regulatory potential" or "bind peak". Furthermore the sentence "rank product [...] vs factor perturbation" was very challenging to follow.
* The description of the implementation in the Methods section seems related but different to the outline of the algorithm.
* As they use "ECDF" graphs as one of only 2 figures, they should be defined early on in the methods. What are they? what do they show? How are they similar to the original? 
* My challenge in judging the validity of the replication is figure 1. From what I can tell these should map to figures 2A and 3A/B in the original manuscript. Unless I am misunderstanding something (is there a stochastic element?) These figures do not match the original manuscript, and the author's "None" line is a straight line whereas it appears to be distinctly non-linear in the original (Is the None line the "static" line?)
* The output from Table 2 also does not match the original with the rank being off by almost an order of magnitude in some cases (relatively) in an absolute sense the error is quite small, is the relative or absolute error important here?
* Also the start and end of the gene do not seem to match. Especially Factor AR the starts and ends are all the same - is this supposed to be?

# Minor Article Comments
* I quite like your concordance measure however it's only an intersection operation. A few words either addressing that the rank is lost, or that the rank is not important would be good here
* Figure 1 and 2 the subplots should be labelled with either (or preferably both) the transcription factor and the target gene as in the original manuscript.
* In the article Figure 1 should describe which original figures it's replication (figure 2A, 3AB?)
* Figure 1 should use red and blue as in the original replication
* Figure 1/2 subpanels should be in the same order as the original 


# Implementation Comments

* The authors provide a docker image however there are no details of the specifics of the implementation: such as R version or operating system, or dependencies. It's also important to note the purely a docker image precludes Windows OS for anyone except those with "pro"
* Their package have a number of undocumented dependencies
    * tidyverse
    * xtable
    * cowplot
    * dplyr
    * broom
* In production code no-longer used, commented out parts should be removed such as 
    * #library(GEOquery) in data-get.R
 It helps to make the code more readable 
* Overall I found the code to be lacking in comments, meaning that piecing together what a piece of code does is incredibly challenging.
For an example: function-def.R is largely undocumented and I do not believe that I fully understand it.
* Lastly I was unable to run any result-generation code as make would often exit with 
`*** No rule to make target 'data/target_output/variants_Tet1_5795.rds', needed by 'article/figures/ecdf.png'.` ( or whichever rule was being run). This may be an issue with make on windows but it seems to think that what's being passed to R is a make rule?

# Summary Comments

The authors have picked an interesting article to implement and a re-implementation would lend support to the original implementation. That said, given what I have understood from their implementation, I do not believe the algorithm has been implemented. As such I recommend the authors fix their figures to match the original as closely as possible. Furthermore, they may wish to consider re-writing their article in a "quasi-pedagogical" form to ensure a reader can understand it without having read the original implementation.

MahShaaban commented:
Thanks @aaronshifman for taking the time to read and comment on the article. I will work on improving the article and the code to address the issues you raised.

delsuc commented:
Thank you very much @aaronshifman  for this comprehensive review.

@brunokieffer any progress on your review ?

MahShaaban committed:
Reply to https://github.com/ReScience/ReScience-submission/pull/59#issuecomment-562890238

    	- Text changes to define terms
	- Link figures to figures in original papar
	- Modified figure 1
	- Document function-def.R

MahShaaban committed:
Merge branch 'AHMED-KIM' of https://github.com/MahShaaban/ReScience-submission into AHMED-KIM

MahShaaban commented:
This is a reply to the issues raised by @aaronshifman 
The changes referred to in this reply are 9907c77a5212a5f4fd29c018eb4656b7377629d9 and the pdf [AHMED-2019.pdf](https://github.com/ReScience/ReScience-submission/files/3942540/AHMED-2019.pdf)


**Major comments**

* We revised the manuscript to define the terms; regulatory potential, binding peaks and factor perturbation.
* The section “implementation of BETA in R” is dedicated to describing the steps of the algorithm as it is implemented in R. We tried to use the existing data structures and functional features of R to build a faster and more flexible implementations. The grouping of steps in functions and their order may therefore differ from the python implementation. This is probably the reason for the discrepancies.
* ECDF was defined in the methods. A description of the specific use of this device was also described in the methods.

Because this was intended to be a replication and not just a reproduction of the original analysis, some differences are to be expected. These differences as they appear in the figures and tables can be attributed to two factors; differences in the  implementation and differences in the analysis.
An example of the differences in implementation is that tied ranks are given different values. In addition, the absolute values of the ranks are different. However, the order of the regions is the same which is the most important in this case. 
In the original analysis there were some decisions such as limiting the analysis to top n peaks and filtering the genes by fold-change > cutoff., which seemed arbitrary. It would be in favor of the robustness of the algorithm if the analysis without these decisions predicted the same TF functions. This was the case in our analysis. The predicted functions of the TFs and the order of the direct target was replicated even though we didn’t limit the peaks or filter out the genes.
* For these reasons, Figure 1 is not identical to the originals. However, the predicted TF functions are the same in all three cases. 
* Similarly, table 2 is not identical to the original. However, the same genes/transcripts were predicted as top direct targets.
* The gene coordinates don’t match because of another implementation difference. In our implementation, genes/transcripts are resized to a regions of interest (200kb in this case). Because we used GRanges the resized regions are the ones kept rather than the original gene/transcript coordinates. This makes no difference in the output since in both cases the regulatory potential is calculated based on the distance to the TSS.

**Minor comments**
* We clarified in the manuscript that the order rather than the absolute values of the rank are the most important.
* Figures were modified to address the rest of the minor comments.

**Implementation comments**
* The details of the implementation are provided here https://github.com/MahShaaban/target. The link is also listed in the manuscript (section “Software environment & reproducibility”).
* The order of the examples as presented in the manuscript makes it easier to present the findings since the cell lines are from human and mouse.
* The docker image, the production code and the Makefile were revised and improved on their documentations.

aaronshifman commented:
@MahShaaban I don't seem to have access to your updated manuscript, can you please upload it?

MahShaaban commented:
I edited my response above to add the pdf of the revised manuscript.

brunokieffer commented:
The manuscrit presented by Mahmoud Ahmed and Deok Ryong Kim propose a implementation in R language of the BETA algorithm initially implemented in Python. The BETA procedure is intended to predict the regulatory role of transcription factors, based on the combined analysis of  Chip-seq and differential expression data. The original BETA algorithm has three purposes : 1- to assign a given factor an activation, repression or neutral role, 2- to infer the factor's target genes and 3- to identify the binding motifs of the factor and its potential coregulators. In the Nature Protocole paper, the approach is illustrated by taking the specific cases of the Androgen Receptor, Estrogen Receptor (ESR1) and Tet1 factors. The reproducibility of the original BETA analysis is assessed by the Empirical Cumulative Distribution Function computed from these three specific genes shown in Figure 1 (as compared to figure 2 of Wang manuscript in Nat Protocole). Although the conclusion drawn from the global shape of the curve is similar in both studies, the definition of the ECDF seems to be different. I suggest that the author use the same definition as in the Wang's paper to ease the comparison of the ranking of regulatory potential scores. In particular, the curve computed for down-regulated AR genes is clearly different, which is puzzling to me since the difference is not commented by the authors. To assess the similarity of the results obtained from both implementations, I would suggest the use of a non-parametric test such as Mann-Whitney to compare the ranks obtained from both programs. Another puzzling point is found in table 2 where the genome coordinates do not match on chr19. The location of the KLK2 gene extends over 200 kb, which is too imprecise to compare with the corresponding data shown in the Nat Protocole paper. Furthermore, the Rank column shown a fractional number which can't be true (a rank is defined as an integer number).
Overall, the presentation of the reproduced results would desserve a large improvement, notably by providing a precise definition of the measurables and entities computed by the R implementation.

MahShaaban commented:
Thanks @brunokieffer for your comments.
The review raises issues regarding the differences between the results of the R and python implementations. These are
1. The shape of the ECDF curves in Figure 1 and Figure 2 (of the original manuscript).
2. The genomic coordinates in table 2 and the original text.
3. The values of the ranks

There are two sources to which these difference could be attributed. First, the specifics of R implementation are not identical to the original since we sat to implement the procedure described in the paper itself not replicating the python code. Related to that is the difference in the data structures and the (intermediary/final) outputs exposed to the users. Second, we used the datasets and annotations but didn't limit the analysis to the decisions made by the authors in the original paper. In a robust replication the original results should hold even when using different analysis decisions such as filtering regions of interest or limiting the peak set to top n peaks. 

We would still argue that the described procedure was faithfully implemented in R and the main results of the original analysis were replicated. The basis for this claim is that; the predicted functions of the transcription factors and the order of the targets are the same. We further demonstrated the robustness of algorithm under varying inputs and cutoff choices. Now, we address the specific points in the review.

1. In the original article, the authors chose to filter out the low differentially expressed genes and to limit the analysis to the top 10000 peaks. Moreover, the procedure for dividing of targets into three groups was not described in the paper. These explain the differences between Figure 1 and Figure 2 of the original paper. The finding of the same predicted transcription functions despite no arbitrary filtering of inputs supports a successful replication.
2. In the R implementation genes/transcripts are resized to a distance of 200kb around the TSS (regions of interest). The resized regions are the ones kept in the subsequent analysis steps and the final output. This is why table 2 may seem different from the text of the original paper. This approach is more transparent since it exposes to the user the true genomic coordinates on which the calculations are based. The gene/transcript coordinates are originally provided by the users so we don't think this would be confused with the returned regions of interest.
3. To calculate the rank product the ranks of the regulatory potential and the signed statistics are divided by the total number of gene. That's probably why there are fractions in the final ranks in both the original paper and the current manuscript. The absolute values of the rank products between the two implementation are different but the order is preserved which is sufficient for ranking the targets.

rougier commented:
@delsuc Gentle reminder

delsuc commented:
Thank you @aaronshifman and @brunokieffer for the reviews, and @MahShaaban for your answers.

@MahShaaban, in this submission, you are replicating the transcriptome and ChIP-seq analysis performed by the BETA program. 
Here, you chose to rewrite a program by changing the language from python to R, ( a good thing, as the original work in a version of python which is now obsolete) and also to modify the algorithm in several important manners.

However both reviewers raise concerns about the replication of the results from the original work, and question the quality of the replication.

For instance:

> @aaronshifman  "...I do not believe the algorithm has been implemented..."

> @brunokieffer  "...the presentation of the reproduced results would deserve a large improvement..."

You are providing long answers to both, with arguments about the modifications in the implementation. However, you did little modifications to the manuscript itself. In particular the modifications of the code, as well as the rational for these modifications is not present in the final manuscript. In addition, you chose to present the results in a different manners from the original work, which to the least makes the comparison difficult.

I suggest you modify the text to better outline the differences from the original work, and the rational for these modifications in order to fully answers the reviewers' concerns. Also, the results should be presented in a way closer to the original work.
For instance:

- Redraw Figure 1 and/or explain the striking differences with fig2a and fig3 from the original work
- present table 2 in way closer to the different tables in the manuscript 

Also, please answer @aaronshifman concern about the implementation details and state in the manuscript the dependencies and the versions of the various packages which are used.

MahShaaban commented:
Thanks @delsuc 

In response to the reviewers, I made changes to the text and the replication code. The code changes were mostly for improving the documentation and therefore didn't lead to further changes in the text. 
The reviewers raised concerns about the differences between the figures and table in the replication and the original article. I tried to provide justifications for the differences and show how they don't change the conclusion. I would be interested to know what the reviewers think of these responses and whether they are satisfied.
As @delsuc suggested, I added a new section to the manuscript to describe the discrepancies between the two articles and their sources. 

[AHMED-2019.pdf](https://github.com/ReScience/ReScience-submission/files/4122376/AHMED-2019.pdf)

delsuc commented:
@aaronshifman @brunokieffer can you please comment on the current state of the manuscript. Does this version answer to all your remarks in your review ? Do you consider this publication ready ?
I have the feeling that certain points remain unanswered.

rougier commented:
Gentle reminder

delsuc commented:
@MahShaaban the current version of the manuscript does not reach a level sufficient for publication.
Are considering to answer all the points raised by the referees ?
In the absence of an answer from you, I think I will have to reject the submission.

MahShaaban commented:
Thanks @delsuc I am certainly willing to modify the submission to meet the reviewers satisfaction.
In the latest version, I modified the text as recommended and replied to the reviewers' questions. I was counting on getting a feedback on these modifications to know which points remain unanswered. 
I thank you all for your patience and useful comments and suggestions

aaronshifman commented:
@delsuc @MahShaaban I'll try to get my review back in the next few days.

@MahShaaban what I will say in the interest of expediency is that it's incredibly difficult to review a replication when the results are exceptionally different. You've made a good effort at explaining these differences in the updated manuscript. These difference often included your changing of parameters or implementation details which as you argue is of a benefit.

This being said it's very difficult to determine if you've accurately implemented the method as you've intentionally changed the results. I can see two ways of mitigating this. Both options require you to implement the original results. Option a would be to, in a supplemental section, reproduce the results in a manner you describe as better and provide and argument. Option b would be to describe the original implementation and your updated implementation in the same place - an option I suspect would be more cumbersome.

I believe that your goal as someone publishing in rescience is to replicate the original manuscript.  Significant arguments about the validity of the original implementation and changer therein are probably outside the scope of the journal and at the very least I am most certainly too far away field wise to review those - @delsuc your input there would be appreciated.

delsuc commented:
Dear @MahShaaban , I agree with @aaronshifman , the whole idea of the ReScience project is to reproduce previous results in an alternative implementation. The fact that the implementation allows improvements over the initial work is certainly a valuable advantage, but the replication step cannot be ignored.
In your case, I believe option a) proposed by @aaronshifman is fine, as a new section in you manuscript, ideally presented before your own improvements. Reproducing figures and tables, in format close to the initial work is also required, so that the reader may easily perceive what is due to the implementation details (such a random seeds) and what is due to the core algorithm.
In addition, I will second @aaronshifman on the fact that a deep discussion on the validity of the original work and comparison with improvements of yours is certainly outside the scope of the journal as I understand it.

MahShaaban committed:
calculate ecdf with top 500 genes

MahShaaban commented:
Thanks @delsuc and @aaronshifman I really appreciate your time and effort. Your comments are helping me a great deal.

In addition to the the reasons I gave earlier for the apparent differences between the figures and the table in the two manuscripts, I'd like to make a few points.

1. I'd like to stress that the differences in "implementation" are not intended as improvements nor as deliberate modifications of the algorithm. In many cases, the differences are merely a result of the fact that things are done differently in R and python. That said, there are differences in the "presentation" of the results or the analysis "parameters" which are not exactly related to the implementation. These are important to emphasis to make evaluating the replication possible.
2. In an updated version of the manuscript ([AHMED-2019.pdf](https://github.com/ReScience/ReScience-submission/files/4369208/AHMED-2019.pdf)), I used the top 500 regulated genes to make groups of regulated genes and calculate the ECDFs. This is to replicate exactly the initial analysis as presented in the BETA paper.
3. I tried to explain why the ECDF curves look different from the original figure. One thing I have not addressed before is, in the original paper, the curves don't reach up to 1, which is probably why the curves look different in the first place. otherwise they are similar in terms of the ordering and the distances between the curves of different groups. To my knowledge the ECDF curve should start at 0 and end at 1. I tried with no success to draw the figures (base R and ggplot2) so that the cumulative fraction is calculated in respect to the total number of genes not to the group members. Your advice regarding this point would be appreciated.

rougier commented:
Gentle reminder

MahShaaban commented:
May I ask if there is any updates coming on this issue? Thanks.

delsuc commented:
Dear @MahShaaban ,

As you obviously chose not to follow the routes proposed by @aaronshifman on March 16th and that I supported on March 19th I really thought you had abandoned the idea of finalizing this submission, and wanted to leave it "as is".

The latest modifications in the manuscript dated March 23rd are quite explicitly pointing out the differences between the original work and yours. However you are missing the point of a replication work. As was pointing out @aaronshifman , correcting/improving a previous work is certainly always valuable work, but is actually outside the scope of the ReScience journal, which is replicating a previous published work, in the purpose of improving its availability. In this respect, it is of great importance that the manuscript reproduces as closely as possible the data and artwork of the original publication. This means not only following the notations and graphical choice of the original authors, their data notations, and their choices as closely as possible. You have not yet convinced the reviewers and this editor that is the case, and you last modification did not change your presentation, but rather added some explanations to the text itself.

If you still wish to improve this submission, which I certainly find worthwhile doing, I renew here my proposal of following @aaronshifman 's option a/, which means first replicating as closely as possible the original work and putting this replication in the first part of the manuscript, and in a second part, either in the text or as proposed in appendices, the current results explaining your different implementation choices and how it impacts the quality or the robustness of the analysis.

First replicating the eventual un-optimal choices of an original work is of key importance, as it allows to directly compare the original work with a potentially improved one, with the code at hand, and the ability to act upon it, thus clearly delineating the effect of any modifications. 

Looking forward to your decision.

MahShaaban commented:
Thanks @delsuc for the detailed reply. I am certainly willing to see this replication through.
My understanding is "[Replication of a scientific study (computational or other) means repeating a published protocol, respecting its spirit and intentions but varying the technical details.](http://rescience.github.io/faq/)" And that "[Replication explores which details matter for reaching a specific scientific conclusion.](http://rescience.github.io/faq/)" Excuse me if I am mistakingly putting more emphasis on certain parts since this is my first attempt at a formal replication. I am using the Rescience guidance as well as yours to implement these ideas in a study.
If I understand correctly, the reviewers are asking for a faithful reproduction of the original results first. Then something like what the current version of the manuscript present would follow. I think I can make this reproduction using the data and the python code from the original work and present it in a section at the beginning of the Results. Would this be a proper way to proceed?

delsuc commented:
Dear @MahShaaban, *"using the data and the python code from the original work"* I believe this is not really the spirit of ReScience, which would rather to adapt your R code to replicate a much as possible the original work, making possible to compare how your proposition differs from the original work. In addition, the matter here is not only the code, but also the way the results are presented (figures and table) which are different from the original work, and make it difficult to compare.

I call the journal Editors on this matter to make a decision on this. - poke @rougier @khinsen @oliviaguest @benoit-girard -

rougier commented:
I fully support @delsuc analysis and recommendation. The idea behind replication (not reproduction) is to provide an alternative implementation based on the article description (and looking at original code only if necessary and available). In your case and as @delsuc suggested, you would need first to provide the R implementation and show how it can reproduce the original results, sticking to the original format as closely as possible to ease the comparison. 

Note also that if you modified and improved the original algorithm, you can consider submitting this new and original part to a "regular" journal and publish only the replication part in ReScience C.

khinsen commented:
I agree with @delsuc and @rougier: the core of a replication article is to show how close you can get to the results of the original paper by using an alternative implementation of the method. If you intentionally deviate from the published method, you are not doing a replication. If different results are due to implementation details (e.g. different data structures in R vs. Python), then you end up with a failed replication, or at least a not completely successful replication. Which is fine, of course, that's the reality of computational science. It's up to the reviewers to decide if the differences in results are important enough to call a replication a failure, because that's a judgment that requires domain expertise.

Going beyond the basic replication by varying other aspects of the method is great - as @rougier said, you could even publish this in "standard" journals, but you are welcome to add this to a ReScience submission as well. But replication needs to be a first step, because it's the only way to figure out if differences come from implementation details or from intentional modifications of the protocol.

MahShaaban commented:
I tried once again, unsuccessfully, to resolve the differences between the two implementations of BETA in R and python. In the initial submission, I used several analysis decisions that were not exactly the same as the original publication. I later modified the submission and used the exact parameters from the original paper, however a few differences persisted. Those were the absolute values of the ranking of the genes (Table) and the shape of the ECDF functions (Figure). I thought that as the order of the ranks and the position of the ECDF functions were the same, the analyses converged on the same conclusions. Based on this, I wrote in the submission that the replication was successful. I proceeded to add a section to the replication explaining the differences in the presentation of the output as suggested here. As the reviewers disagreed, I think the more appropriate conclusion would be a partial or a failed replication.
If the reviewers see a value in the work as it stands, would they be willing to recommend a conclusion of partial or failed replication? Of course, I would be willing to make other changes in the text accordingly.

Thanks

rougier commented:
@MahShaaban Sorry for the long delay.
@delsuc @khinsen @aaronshifman @brunokieffer Re-activating this submission.

From what I've understood, you managed to replicate some results but failed to replicate some other results and from the journal perspective, it 's perfectly ok to publish partial or failed replication as long as there are accompanying explanations.

MahShaaban commented:
I would like to ask if there is any update on this issue. Thanks.

rougier commented:
@delsuc @khinsen @aaronshifman @brunokieffer @MahShaaban

Dear all, I think we need to make a decision for this submission but maybe we need first to have a summary of the status since the thread is already quite long. @delsuc, can you tell if the submission can be accepted and if no, what is missing?

delsuc commented:
Here is how I see the whole story:

After the manuscript submission from @MahShaaban (Sept 2019), reviewers @brunokieffer and @aaronshifman did a fairly comprehensive work on reviewing this work (Dec 2019).
Both reviewers pointed out both the interest of the work and the discrepancies between the target work and this one, particularly in the algorithms. Besides, the pictures and data tables chosen by @MahShaaban do not match the figures from the original work, and make it difficult to compare both results. Both reviewers asked for substantial modifications in order to correct these weaknesses.

From these review, the author explained his choice and developed his approach in this thread (Dec 2019)

As editor, I asked the author (Jan 2020) to answer the main points raised by the reviewer, not only as a direct answer to them, but also by adapting the manuscript to their main concerns.
A revision of the manuscript was submitted by the author (Jan 2020) with a more detailed description of the departure of this work from the original one, but still lacking a real replication of the original work, and no modification in the artworks and tables.

A discussion has been engaged in spring 2020 between the author, the reviewers and myself, whether the manuscript in this state was reaching the requirement for being published in ReScience. We asked @MahShaaban to revise the submission to more closely match the original work, eventually leaving his modifications/improvements in a second part of the manuscript, providing the reader to a valuable comparison with this improved alternative. 
Confronted with some difficulties and misunderstanding I called to the editors for arbitrage, and this decision was supported by the board (4 Aug 2020).

Since then, to my knowledge, no action was taken by the author, unless I missed some important steps, which is not unlikely, given my lack of experience with the details of the github processes.

MahShaaban commented:
Thanks @delsuc I think this is a fair description of the thread.
I would add that On Oct 2, 2020, I made this https://github.com/ReScience/ReScience-submission/pull/59#issuecomment-702675463.
Very briefly, I think I followed the description of the algorithm as, it occurs in the original paper, faithfully. The R implementation gave results that differed numerically from the original (the shape of the distribution functions, for example), but didn't change the conclusions (the aggregate functions or the ordering of the gene targets).

So, I am asking
> If the reviewers see a value in the work as it stands, would they be willing to recommend a conclusion of partial or failed replication? Of course, I would be willing to make other changes in the text accordingly.

rougier commented:
I think the work is valuable but need to be improved. From re-reading all the exchange here, I think it is necessary to explain why you do not get the same result. You explained this is due to difference betwen R and Python but I think more details are needed because the two langages are equivalent and one thing that can be done in Python can be replicated in R. I can only think on a few low-level corner cases where this might be difficult. Since you have access to the Python code, you try a literal transcription of the Python code to R code to check where is the differrence. I just did that myself (from Matlab to Python) because the description of the algorithm in the paper did not lead to the same results. After having written the litteral transciption, we get the same results and identified the problem in one random matrix initialisation. I know this is extra work for your, but this is precisely the kind of information that we look for in ReScience. Later, if someone read the original article and your article with this supplementary information, they will know how to rewrite it using any language.

If you don't find where the difference comes from, you ca also contact the original authors to ask for some information. This is also something I do regularly and authors are generally happy to answer.

MahShaaban commented:
Thanks, @rougier I have looked into python code that generates the figures. I tried to replicate the same procedure in R but still couldn't find where the difference is.
I contacted the author with a few questions before I opened this issue but unfortunately, I didn't get a response.

rougier commented:
First thing is to run authors version of their script to check you can get their results from their code. If this is not the case, there might be a problem.

MahShaaban commented:
Thanks @rougier I did that and the figures agreed with what's presented in the protocol paper.

rougier commented:
So next step would be to do a "literal" transcription of these sources into python to check if you obtain the same result. This means you have to control at each step the content of variable is exactly the same. This is good way to check (for example) if they use double or float.
